// @auto/img-crop/dist/cjs/img-crop.development.js
'use strict';

// 通用图片裁切
// https://git.corpautohome.com/gp_rd_fe/tools/tech/utils/tree/master/packages/img-crop#readme

/* eslint-disable no-multi-spaces, consistent-return */
var $host = '(\\.autoimg\\.cn/)';
var $path = '(/g\\d+/M\\d+[A-F]?/[\\dA-z]+/[\\dA-z]+/)'; // 目录

var $rule = '(.*autohome[^_]*__)?'; // 裁切规则

var $suffix = 'autohomecar__'; // 裁切规则后缀

var $name = '(.+\\.(jpe?g|png|gif|bmp))'; // 文件名
// dis 2.0 裁切（松散规则）

function rule_dis(src, rule) {
  var $1 = '(//.+)'; // '//' + 子域名（保留）

  var $2 = $host; // .autoimg.cn/（保留）

  var $3 = '(.*)'; // 存储目录（保留）

  var $4 = $path; // 存储目录（保留）

  var $5 = $rule; // 裁切规则（删除）

  var $6 = $name; // 图片名（保留）

  var reg = new RegExp("".concat($1).concat($2).concat($3).concat($4).concat($5).concat($6), 'i');

  if (reg.test(src)) {
    return src.replace(reg, "$1$2$3$4".concat(rule).concat($suffix, "$6"));
  }
} // 养车 yc[01]\.autoimg.cn 只能裁切固定的几个尺寸：
// 124x180_0_q30_ 186x140_0_q75_ 240x180_0_q30_
// 详见 http://192.168.100.11/rulelist


function rule_yc(src, rule) {
  var $1 = '(//yc[01])'; // '//' + 子域名（保留）

  var $2 = $host; // .autoimg.cn/（保留）

  var $3 = '(imgs/yc/[\\w/]+/)'; // 存储目录（保留）

  var $4 = '(.+_)?'; // 裁切规则（删除）

  var $5 = $name; // 图片名（保留）

  var reg = new RegExp("".concat($1).concat($2).concat($3).concat($4).concat($5), 'i');

  if (reg.test(src)) {
    return src.replace(reg, "$1$2$3".concat(rule, "$5"));
  }
} // 二手车 2sc2.autoimg.cn/escimg


function rule_2sc(src, rule) {
  var reg = /\/\/2sc2/;

  if (reg.test(src)) {
    return rule_dis(src.replace(reg, '//www2'), rule);
  }
} // 论坛 club2.autoimg.cn/album


function rule_club(src, rule) {
  var $1 = '(//)'; // 协议后面的两个 '//'（保留）

  var $2 = '(.+)'; // 子域名（需要替换为 m1）

  var $3 = $host; // .autoimg.cn/（保留）

  var $4 = '(album/?)?'; // 一级目录（需要替换为 clubdfs）

  var $5 = $path.replace('/g', '/?g'); // 存储目录（保留）

  var $6 = '(userphotos/\\d+/\\d+/\\d+/\\d+/)?'; // 老存储目录（删除）

  var $7 = '(ssq_|\\d+_|jh\\d+_)?'; // 老裁切规则（删除）

  var $8 = $name; // 图片名（保留）

  var reg = new RegExp("".concat($1).concat($2).concat($3).concat($4).concat($5).concat($6).concat($7).concat($8), 'i');

  if (reg.test(src)) {
    return src.replace(reg, "$1m1$3clubdfs/$5".concat(rule).concat($suffix, "$8"));
  }
} // 产品库 car[01]\.(\.m)?autoimg.cn


function rule_car(src, rule) {
  var $1 = '(//)'; // 协议后面的两个 '//'（保留）

  var $2 = '(car[01])'; // 子域名（需要替换为 m0）

  var $3 = '(\\.m)?'; // 子域名（删除）

  var $4 = $host; // .autoimg.cn/（保留）

  var $5 = '([\\w/]+/)'; // 存储目录（保留）

  var $6 = '(.+_)?'; // 老裁切规则（删除）

  var $7 = $name; // 图片名（保留）

  var reg = new RegExp("".concat($1).concat($2).concat($3).concat($4).concat($5).concat($6).concat($7), 'i');

  if (reg.test(src)) {
    return src.replace(reg, "$1m0$4cardfs/$5".concat(rule).concat($suffix, "$7"));
  }
} // 游记/旅行家


function rule_you(src, rule) {
  var $1 = '(//)'; // 协议后面的两个 '//'（保留）

  var $2 = '(img[23])'; // 子域名（需要替换为 m1）

  var $3 = $host; // .autoimg.cn/（保留）

  var $4 = '(travelplat|pano)'; // 存储目录（保留）

  var $5 = $path; // 存储目录（保留）

  var $6 = $rule; // 裁切规则（删除）

  var $7 = $name; // 图片名（保留）

  var reg = new RegExp("".concat($1).concat($2).concat($3).concat($4).concat($5).concat($6).concat($7), 'i');

  if (reg.test(src)) {
    return src.replace(reg, "$1m1$3$4$5".concat(rule).concat($suffix, "$7"));
  }
}

function checkWebp() {
  try {
    return document.createElement('canvas').toDataURL('image/webp').indexOf('data:image/webp') === 0;
  } catch (err) {
    return false;
  }
}

var isSupportWebp; // 当前环境是否支持 webp
// 七牛云 & 车家号

function rule_qiniu(src, rule) {
  var reg1 = /\?image(View|Mogr)2.+/;
  var $1 = "(//qnwww[23]".concat($host, "youchuang").concat($path).concat($name, ")");
  var reg2 = new RegExp("".concat($1, "(\\?imageView2.+|\\?imageMogr2.+)?"), 'i');

  if (reg1.test(src) || reg2.test(src)) {
    var newRule = rule.replace(/(\d+)x(\d+)(.+q(\d+).+)?/, function (r, w, h, r2) {
      var q = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : 87;

      if (isSupportWebp === undefined) {
        isSupportWebp = checkWebp();
      }

      var format = "format/".concat(isSupportWebp ? 'webp' : "jpg/q/".concat(q), "/");
      return "?imageView2/1/interlace/1/w/".concat(w, "/h/").concat(h, "/").concat(format);
    });

    if (reg1.test(src)) {
      return src.replace(reg1, "".concat(newRule));
    }

    return src.replace(reg2, "$1".concat(newRule));
  }
}

function universal(src, rule) {
  return rule_2sc(src, rule) // 二手车
  || rule_club(src, rule) // 论坛
  || rule_car(src, rule) // 产品库
  || rule_you(src, rule) // 游记/旅行家
  || rule_yc(src, rule) // 养车
  || rule_qiniu(src, rule) // 七牛云 & 车家号
  || rule_dis(src, rule) // dis通用裁切
  ;
}
/**
 * 图片裁切
 *
 * @param {string} src 图片地址
 * @param {string} rule 裁切规则
 * @returns {string} 裁切后的图片地址
 */


function imgCrop(src, rule) {
  // 只处理 autoimg.cn 域
  if (!new RegExp("//.+".concat($host, ".*").concat($name), 'i').test(src)) {
    return src;
  }

  var newSrc = universal(src, rule);
  return (newSrc || src).replace(/https?:/, '');
}
/**
 * 在 iPad 中使用多倍尺寸
 *
 * @param {string} rule 裁切规则
 * @param {number} scale 缩放比例, 默认 2.56
 * @returns {string} 新裁切规则
 */


function imgScale(rule) {
  var scale = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 2.56;

  if (!/iPad/i.test(navigator.userAgent)) {
    return rule;
  }

  return rule.replace(/(\d+)x(\d+)/, function (r, width, height) {
    return "".concat(Math.ceil(width * scale), "x").concat(Math.ceil(height * scale));
  });
}
/**
 * 获取资讯图片地址 (16年以后都有16:9图)
 *
 * @param {Object} data                - 含有多个图片地址的 Object
 * @param {string} data.AppImg         - APP 代表图 16:9(560*315)
 * @param {string} data.FirstCoverImg  - APP 封面图1 4:3(560*420) 接口实际输出16:9
 * @param {string} data.SecondCoverImg - APP 封面图2 4:3(560*420) 接口实际输出16:9
 * @param {string} data.ThirdCoverImg  - APP 封面图3 4:3(560*420) 接口实际输出16:9
 * @param {string} data.picUrls        - PC/M 代表图 4:3(400*300)
 * @returns {string} 图片地址
 */


function getImgSrc(data) {
  var isDefault = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
  return data.picUrls || data.AppImg || data.FirstCoverImg || data.SecondCoverImg || data.ThirdCoverImg || (isDefault ? '//s.autoimg.cn/sou/soum/images/default.png' : '');
}

exports.checkWebp = checkWebp;
exports.getImgSrc = getImgSrc;
exports.imgCrop = imgCrop;
exports.imgScale = imgScale;
